File: gridworld/gridworld.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.table import Table
from IPython import embed

DISCOUNT = 1
GRID_SIZE = 4
GRID = np.arange(GRID_SIZE*GRID_SIZE).reshape(GRID_SIZE, GRID_SIZE)
ACTIONS = [
            np.array([0, -1]), #left
            np.array([0,  1]), #right
            np.array([-1, 0]), #up
            np.array([1,  0])  #down
]
ACTION_PROB = 0.25
           
def step(state, action):
    if state == [0, 0]:
        return [0, 0], 0.0
    if state == [3, 3]:
        return [3, 3], 0.0

    next_state = (np.array(state) + action).tolist()
    x, y = next_state
    if x < 0 or x >= GRID_SIZE or y < 0 or y >= GRID_SIZE:
        reward = -1.0
        next_state = state
    else:
        reward = -1.0
    return next_state, reward

def show_grid(grid):
    fig, ax = plt.subplots()
    ax.set_axis_off()
    tb = Table(ax, bbox=[0, 0, 1, 1])

    nrows, ncols = grid.shape
    width, height = 1.0/ncols, 1.0/nrows

    for (i, j), val in np.ndenumerate(grid):
        color = 'grey' if (i, j) == (0, 0) or (i, j) == (3, 3) else 'white'
        tb.add_cell(i, j, width, height, text=val,
                    loc='center', facecolor=color)

    ax.add_table(tb)
    plt.show()

value = np.zeros_like(GRID, dtype=np.float64)

for _ in range(10):
    new_value = np.zeros_like(GRID, dtype=np.float64)
    for (i, j) in np.ndindex(GRID.shape):
        for action in ACTIONS:
            (next_i, next_j), reward = step([i, j], action)
            new_value[i, j] += ACTION_PROB * (reward + DISCOUNT * value[next_i, next_j])
    value = new_value
    show_grid(value)

File: gridworld/gridworld.jl
using Plots
using LinearAlgebra

const γ = 1
const SIZE = 4
const GRID = reshape(1:SIZE*SIZE, (SIZE, SIZE)) 

const ACTIONS = [
                 CartesianIndex(0, -1), # left
                 CartesianIndex(0, 1),  # right
                 CartesianIndex(-1, 0), # up
                 CartesianIndex(1,0)    # down
                ]

const ACTION_PROB = 0.25

const n_actions = 4
const n_states = 16

isterminal(s::CartesianIndex{2}) = 
        s == CartesianIndex(1,1) || s == CartesianIndex(4,4) 

function step(s::CartesianIndex{2}, a::CartesianIndex{2})
    ns = s + a
    if isterminal(s) || ns[1] < 1 || ns[1] > SIZE || ns[2] < 1 || ns[2] > SIZE
        ns = s
    end
    r = isterminal(s) ? 0.0 : -1.0

    (ns, r)
end

function PolicyEval!(value::Array{Float64, 2}, iters=100)
    for _ = 1:iters
        new_value = zeros(Float64, SIZE, SIZE)
        for ind in CartesianIndices((SIZE, SIZE))
            for action in ACTIONS
                next_ind, reward = step(ind, action)
                new_value[ind] += ACTION_PROB * (reward + γ * value[next_ind])
            end
        end
        value .= new_value
    end
    value
end

#function PolicyImprove(value::Array{Float64, 2})
#    policy = zeros(Float64, SIZE, SIZE)
#    for ind in CartesianIndices((SIZE, SIZE))

value = zeros(Float64, SIZE, SIZE)
PolicyEval!(value)
display(value); println()

#i = CartesianIndex(2, 2)
#for action in ACTIONS
#    println(action)
#    j, r = step(i, action)
#    println(j, r + γ * value[j])
#end

π = fill(0.25, n_states, n_actions)

println("random policy")
display(π)

linear = LinearIndices((SIZE, SIZE))
cartesian = CartesianIndices((SIZE, SIZE))

for s in linear
    chosen_a = argmax(π[s])

    action_vals = zeros(n_actions)
    for a in 1:n_actions
        (ns, r) = step(cartesian[s], ACTIONS[a])
        action_vals[a] = r + γ * value[ns]
    end

    best_a = argmax(action_vals)

    if chosen_a != best_a
        π[s, :] = Matrix{Float64}(I, n_actions, n_actions)[best_a, :]
    end
end

println("updated policy")
display(π)
